angular/src/components/set-pin.component.ts

```diff
@@ -0,0 +1,49 @@
+import { Directive } from '@angular/core';
+
+import { CryptoService } from 'jslib-common/abstractions/crypto.service';
+import { StorageService } from 'jslib-common/abstractions/storage.service';
+import { UserService } from 'jslib-common/abstractions/user.service';
+import { VaultTimeoutService } from 'jslib-common/abstractions/vaultTimeout.service';
+
+import { ConstantsService } from 'jslib-common/services/constants.service';
+
+import { Utils } from 'jslib-common/misc/utils';
+
+import { ModalRef } from './modal/modal.ref';
+
+@Directive()
+export class SetPinComponent {
+
+    pin = '';
+    showPin = false;
+    masterPassOnRestart = true;
+
+    constructor(private modalRef: ModalRef, private cryptoService: CryptoService, private userService: UserService,
+        private storageService: StorageService, private vaultTimeoutService: VaultTimeoutService) { }
+
+    toggleVisibility() {
+        this.showPin = !this.showPin;
+    }
+
+    async submit() {
+        if (Utils.isNullOrWhitespace(this.pin)) {
+            this.modalRef.close(false);
+        }
+
+        const kdf = await this.userService.getKdf();
+        const kdfIterations = await this.userService.getKdfIterations();
+        const email = await this.userService.getEmail();
+        const pinKey = await this.cryptoService.makePinKey(this.pin, email, kdf, kdfIterations);
+        const key = await this.cryptoService.getKey();
+        const pinProtectedKey = await this.cryptoService.encrypt(key.key, pinKey);
```

Comment:
I know this is a bit outside the scope of this PR, but my personal opinion is that this code should belong to a service. Currently this component needs to know implementation designs to retrieve keys and do cryptographic operations. I ponder if this anti pattern originates from the key being retrievable from `CryptoService`.